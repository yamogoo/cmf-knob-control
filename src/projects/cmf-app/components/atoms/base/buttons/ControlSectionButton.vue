<template lang="pug">
div(
    ref="refRoot"
    :class="[CLASS_NAME, {'focused': isFocused}, isPressed ? Statuses.PRESSED : Statuses.RELEASED]"
    @mousedown="onMouseDown"
)
    slot
</template>

<script setup lang="ts">
import { onUnmounted, ref, watch } from 'vue';
import { APP_PREFIX } from '@app/config';
import g from 'gsap';

const CLASS_NAME = `${APP_PREFIX}-control-boxed-button`;

interface Props {
    id: number,
    isKnobPressed: boolean,
    isFocused: boolean
}

const props = withDefaults(defineProps<Props>(), {});

const emit = defineEmits<{
    (e: 'onPress', id: number): void
}>();

const refRoot = ref<HTMLDivElement | null>(null);
const isPressed = ref(false);

const onMouseDown = (e: MouseEvent): void => {
    e.preventDefault();
    // e.stopPropagation();

    onPress(true);

    if (refRoot.value) {
        refRoot.value.addEventListener('mouseup', mouseUpHandler);
        refRoot.value.addEventListener('mouseout', mouseOutHandler);
        refRoot.value.addEventListener('mousemove', mouseMoveHandler);
    };
};

watch(() => props.isKnobPressed, () => {
    isPressed.value = props.isKnobPressed;
    onAnimPress(isPressed.value);
});

const mouseMoveHandler = (_e: MouseEvent) => {

};

const mouseUpHandler = (e: MouseEvent) => {
    e.preventDefault();
    // e.stopPropagation();

    onPress(false);

    emit('onPress', props.id);
};

const mouseOutHandler = (e: MouseEvent) => {
    e.preventDefault();
    // e.stopPropagation();

    onPress(false);

    if (refRoot.value) {
        refRoot.value.removeEventListener('mouseup', mouseUpHandler);
        refRoot.value.removeEventListener('mouseout', mouseOutHandler);
    };
};

onUnmounted(() => {
    onPress(false);

    if (refRoot.value) {
        refRoot.value.removeEventListener('mouseup', mouseUpHandler);
        refRoot.value.removeEventListener('mouseout', mouseOutHandler);
    };
});

const onPress = (state: boolean) => {
    isPressed.value = state;

    onAnimPress(isPressed.value);
};

/* * * Animations * * */

const onAnimPress = (state: boolean): void => {
    const scale = state ? .9 : 1.;

    if (refRoot.value) {
        g.to(refRoot.value, {
            scale,
            duration: .25,
            ease: 'power4.out'
        });
    };
};

</script>

<script lang="ts">
export enum Classes {

}

export enum Statuses {
    PRESSED = 'pressed',
    RELEASED = 'released'
}
</script>

<style lang="scss">
.#{$APP_PREFIX}-control-boxed-button {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100%;
    border-radius: 24px;
    @include transition(color background-color, 25ms, ease-out);
    color: $GRAY;

    &.released {
        background-color: $GRAY_SHADOW;
    }

    &.focused {
        // background-color: lighten($GRAY_SHADOW, 9%);
        color: lighten($CARDAMON_FADE, 9%);
        background-color: lighten($CARDAMON, 9%);
    }

    &.pressed {
        background-color: lighten($GRAY_SHADOW, 21%);
    }
}
</style>