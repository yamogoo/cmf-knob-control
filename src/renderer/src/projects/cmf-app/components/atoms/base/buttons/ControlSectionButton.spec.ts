import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'

import ControlSectionButton from './ControlSectionButton.vue'

describe('ControlSectionButton', () => {
  test.each([1, 2, 3])('should emit "onPress" event with id = (%i)', (id) => {
    const component = mount(ControlSectionButton, {
      props: {
        id,
        isKnobPressed: false,
        isFocused: false
      }
    })

    const buttonEl = component.find('[data-test-id="control-section-button"]')
    buttonEl.trigger('mousedown')
    buttonEl.trigger('mouseup')

    const emmitedId = (component.emitted('onPress') as Array<Array<number>>)[0][0]
    expect(emmitedId).toBe(id)
    expect(component.emitted('onPress')).toMatchSnapshot()
  })

  test.each(['<p>Slot Content</p>'])('should render slot content (%s)', (slotContent) => {
    const component = mount(ControlSectionButton, {
      slots: {
        default: slotContent
      }
    })

    const html = component.find('[data-test-id="control-section-button"]').html()
    expect(html).toContain(slotContent)
    expect(html).toMatchSnapshot()
  })

  test.each(['focused', 'released'])('should have "%s" class', (className) => {
    const component = mount(ControlSectionButton, {
      props: {
        id: 1,
        isFocused: true,
        isKnobPressed: false
      }
    })

    const buttonEl = component.find('[data-test-id="control-section-button"]')

    expect(buttonEl.classes(className)).toBe(true)
    expect(buttonEl.classes()).toMatchSnapshot()
  })

  test.each(['pressed'])(
    'should have "%s" class by passign "isKnobPressed" prop from parent',
    (className) => {
      const component = mount(ControlSectionButton, {
        props: {
          id: 1,
          isFocused: false,
          isKnobPressed: true
        }
      })

      const buttonEl = component.find('[data-test-id="control-section-button"]')

      expect(buttonEl.classes(className)).toBe(true)
      expect(buttonEl.classes()).toMatchSnapshot()
    }
  )

  test.each(['pressed'])('should have "%s" class by pressing the button', (className) => {
    const component = mount(ControlSectionButton, {
      props: {
        id: 1,
        isFocused: false,
        isKnobPressed: true
      }
    })

    const buttonEl = component.find('[data-test-id="control-section-button"]')
    buttonEl.trigger('mousedown')

    expect(buttonEl.classes(className)).toBe(true)
    expect(buttonEl.classes()).toMatchSnapshot()
  })
})
