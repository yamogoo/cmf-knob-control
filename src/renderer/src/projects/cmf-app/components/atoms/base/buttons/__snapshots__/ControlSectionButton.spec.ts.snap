// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ControlSectionButton > should emit "onPress" event with id = (1) 1`] = `
[
  [
    1,
  ],
]
`;

exports[`ControlSectionButton > should emit "onPress" event with id = (2) 1`] = `
[
  [
    2,
  ],
]
`;

exports[`ControlSectionButton > should emit "onPress" event with id = (3) 1`] = `
[
  [
    3,
  ],
]
`;

exports[`ControlSectionButton > should have "focused" class 1`] = `
[
  "cmf-app-control-boxed-button",
  "focused",
  "released",
]
`;

exports[`ControlSectionButton > should have "pressed" class by passign "isKnobPressed" prop from parent 1`] = `
[
  "cmf-app-control-boxed-button",
  "pressed",
]
`;

exports[`ControlSectionButton > should have "pressed" class by pressing the button 1`] = `
[
  "cmf-app-control-boxed-button",
  "pressed",
]
`;

exports[`ControlSectionButton > should have "released" class 1`] = `
[
  "cmf-app-control-boxed-button",
  "focused",
  "released",
]
`;

exports[`ControlSectionButton > should render slot content (<p>Slot Content</p>) 1`] = `
"<div class="cmf-app-control-boxed-button released" data-test-id="control-section-button">
  <p>Slot Content</p>
</div>"
`;
